lib/MetaCPAN/Web/Role/Elasticsearch/Adapter.pm

```diff
@@ -0,0 +1,50 @@
+package MetaCPAN::Web::Role::Elasticsearch::Adapter;
+
+use Moose::Role;
+
+=head1 METHODS
+
+=head2 extract_first_element
+
+Elasticsearch 1.x changed the data structure returned when fields are used.
+For example before one could get a ArrayRef[HashRef[Str]] where now
+that will come in the form of ArrayRef[HashRef[ArrayRef[Str]]]
+
+This function will select the first element given an array of hash of array.
+So this:
+
+    [
+      {
+        distribution => ['WhizzBang'],
+        author       => ['EVE'],
+      },
+       ...
+   ]
+
+becomes:
+
+    [
+      {
+        distribution => 'WhizzBang',
+        author       => 'EVE',
+      },
+      ...
+    ]
+
+=cut
+
+sub extract_first_element {
+    my ( $self, $array ) = @_;
+    foreach my $hash ( @{$array} ) {
+        foreach my $field ( %{$hash} ) {
+            my $value = $hash->{$field};
+
+            # We only extract the first element when have an ArrayRef
+            next if not( ref($value) and ( ref($value) eq 'ARRAY' ) );
```

Comment:
It's a good point you bring up.  Yes it's intended only for single element
arrays where the field value was just a scalar before.   However, I'm open
to adding a count check and die if the value encountered is a multi-element
array?

2015-01-09 10:31 GMT-07:00 Randy Stauner <notifications@github.com>:

> In lib/MetaCPAN/Web/Role/Elasticsearch/Adapter.pm
> <https://github.com/CPAN-API/metacpan-web/pull/1438#discussion-diff-22729578>
> :
>
> > +        distribution => 'WhizzBang',
> > +        author       => 'EVE',
> > +      },
> > +      ...
> > +    ]
> > +
> > +=cut
> > +
> > +sub extract_first_element {
> > +    my ( $self, $array ) = @_;
> > +    foreach my $hash ( @{$array} ) {
> > +        foreach my $field ( %{$hash} ) {
> > +            my $value = $hash->{$field};
> > +
> > +            # We only extract the first element when have an ArrayRef
> > +            next if not( ref($value) and ( ref($value) eq 'ARRAY' ) );
>
> Should this be testing the length of the array?
> You don't want to discard other elements if there are multiple.
> Or is this only intended to be used for certain queries?
> I see that the current places where it's used there shouldn't be any
> multi-element arrays.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CPAN-API/metacpan-web/pull/1438/files#r22729578>.
>